src/template.cpp

```diff
@@ -0,0 +1,146 @@
+#include <Rcpp.h>
+using namespace Rcpp;
+
+// Positive substring: return a substring of `str``going from `start` to `end`;
+// if `end` is earlier than `start`, return empty string.
+std::string pos_substr(std::string str, int start, int end) {
```

Comment:
FWIW, I think it's better to treat `start` as inclusive and `end` as exclusive. This has a number of nice benefits over having an inclusive `end`:

- The length of the substring is just `end - start`
- You can move onto the next piece by using the old `end` value as the new `start` value
- It's possible to get a zero length string by having `start == end`; you generally don't get into `end < start` situations like you do here